### DISCUSSION TITLE
Upload file using rest API : WebDav

### DISCUSSION THREAD

**Discussion by rashmimaharana** [drakkan/sftpgo#538] [Q&A] [answered]
Hi @drakkan and Team,

Do you have any example docs to use rest API to upload/download files. I have enabled WebDav port and can fetch user details, but not sure how to upload. 
Again thanks for addressing all my concerns. :)

FYI :
I am doing some changes to capture email id from end user and displayed the same in client page, will raise a PR for the same.

![image](https://user-images.githubusercontent.com/10846216/133598868-42491ca1-ee2b-42e7-93df-9a7c6951d21a.png)

Thanks,
Rashmi Maharana

**Comment by drakkan** [comment, marked as answer]
REST API and webdav are different things. To use WebDAV you need a client implementing the protocol, for example Nautilus, windows explorer etc..

You can test REST API with any openapi tool. I generally use [swaggerui](https://github.com/swagger-api/swagger-ui). Please note that this is a test tool, REST API should be used within an external integration.

Regarding your changes they look interesting but we actually don't have an email field for users, I plan to add it in future versions (it is useful for password recovery), we likely need a database migration.

**Reply by rashmimaharana** [reply]
Hi @drakkan 

I tried to use one of the users API to upload files, but getting the message : " No files uploaded! " every time. Am i missing anything here. 
Tried with postman and from shell using curl. 

Steps :
1> Get user token first.
API : http://10.125.XX.XXX:8080/api/v2/user/token
Method type : GET
Authorization : Username : test , password : ***
Full curl command taken from postman :
curl --location --request GET 'http://10.125.XX.XXX:8080/api/v2/user/token' \
--header 'Authorization: Basic QVhkW.....'
Response : 
{
    "access_token": "eyJhbGciOiJIUzI1N.......",
    "expires_at": "2021-09-29T12:46:07Z"
}

2> Used the above token to upload file.
API : http://10.125.XX.XXX:8080/api/v2/user/files
Method type : POST

Full curl command taken from postman :
curl --location --request POST 'http://10.125.XX.XXX:8080/api/v2/user/files' \
--header 'Authorization: Bearer eyJhbGciOiJIUzI1N.......' \
--header 'Content-Type: multipart/form-data' \
--form 'test3.txt=@"/root/test.txt"'

Response :
{
    "message": "No files uploaded!"
}
Logs :
{"level":"debug","time":"2021-09-29T09:44:17.050","sender":"HTTP","connection_id":"HTTP_c5a3asddeqhfb2qf9180","message":"connection added, local address \"10.125.XX.XXX:8080\", remote address \"10.33.XXX.XXX:60150\", num open connections: 1"}
{"level":"debug","time":"2021-09-29T09:44:17.328","sender":"HTTP","connection_id":"HTTP_c5a3asddeqhfb2qf9180","message":"connection removed, local address \"10.125.XX.XXX:8080\", remote address \"10.33.XXX.XXX:60150\" close fs error: <nil>, num open connections: 0"}
{"level":"info","time":"2021-09-29T09:44:17.328","sender":"httpd","local_addr":"10.125.XX.XXX:8080","method":"POST","proto":"HTTP/1.1","remote_addr":"10.33.XXX.XXX:60150","request_id":"sftpgo-1-test-kr/YsPpxc17LZ-000204","uri":"http://10.125.XX.XXX:8080/api/v2/user/files","user_agent":"PostmanRuntime/7.28.4","resp_status":400,"resp_size":33,"elapsed_ms":278}

**Reply by drakkan** [reply]
Please check the docs [here](https://sftpgo.stoplight.io/docs/sftpgo/b3A6MTY2OTU4OTM-upload-files). It must be `--form filename=...` not `--form 'test3.txt=@"/root/test.txt"'`

**Reply by rashmimaharana** [reply]
Thanks @drakkan :)
It worked.
